pkg/crypto/certmanager/storage/awssecretsmanager/impl.go

```diff
@@ -0,0 +1,224 @@
+	metadataClient, err := getMetadataClient()
+	if err != nil {
+		return nil, err
+	}
+	var region string
+	if arn, err := arn.Parse(secretId); err == nil {
+		region = arn.Region
+	} else {
+		region, err = metadataClient.Region()
+		if err != nil {
+			return nil, err
+		}
+	}
+	awsSession, err := session.NewSession(&aws.Config{
+		Region: aws.String(region),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error creating session: %s", err)
+	}
+	if awsSession == nil {
+		return nil, errors.New("awsSession == nil")
+	}
+	return secretsmanager.New(awsSession), nil
+}
+
+func getAwsSecret(awsService *secretsmanager.SecretsManager,
+	secretId string) (map[string]string, error) {
+	input := secretsmanager.GetSecretValueInput{SecretId: aws.String(secretId)}
+	output, err := awsService.GetSecretValue(&input)
+	if err != nil {
+		return nil,
+			fmt.Errorf("error calling secretsmanager:GetSecretValue: %s", err)
+	}
+	if output.SecretString == nil {
+		return nil, errors.New("no SecretString in secret")
+	}
+	secret := []byte(*output.SecretString)
+	var secrets map[string]string
+	if err := json.Unmarshal(secret, &secrets); err != nil {
+		return nil, fmt.Errorf("error unmarshaling secret: %s", err)
+	}
+	return secrets, nil
+}
+
+func newLS(secretId string, logger log.DebugLogger) (*LockingStorer, error) {
+	awsService, err := getAwsService(secretId)
+	if err != nil {
+		return nil, err
+	}
+	return &LockingStorer{
+		awsService: awsService,
+		logger:     logger,
+		secretId:   secretId,
+	}, nil
+}
+
+func putAwsSecret(awsService *secretsmanager.SecretsManager,
+	secretId string, secrets map[string]string) error {
+	secretString, err := json.Marshal(secrets)
+	if err != nil {
+		return err
+	}
+	input := secretsmanager.PutSecretValueInput{
+		SecretId:     aws.String(secretId),
+		SecretString: aws.String(string(secretString)),
+	}
+	output, err := awsService.PutSecretValue(&input)
+	if err != nil {
+		return fmt.Errorf("error calling secretsmanager:PutSecretValue: %s",
+			err)
+	}
+	for _, versionStage := range output.VersionStages {
+		if *versionStage == "AWSCURRENT" {
+			return nil
+		}
+	}
+	return errors.New("no AWSCURRENT version stage associated")
+}
+
+func (ls *LockingStorer) read() (*certmanager.Certificate, error) {
+	keyMap, err := getAwsSecret(ls.awsService, ls.secretId)
+	if err != nil {
+		return nil, err
+	}
+	certPEM := &bytes.Buffer{}
+	for index := 0; ; index++ {
+		certificateBase64 := keyMap[fmt.Sprintf("Certificate%d", index)]
+		if certificateBase64 == "" {
+			if index == 0 {
+				return nil, errors.New("no Certificate in map")
+			}
+			break // We've reached the end of the certificate chain.
+		}
+		certDER, err := base64.StdEncoding.DecodeString(
+			strings.Replace(certificateBase64, " ", "", -1))
+		if err != nil {
+			return nil, err
+		}
+		if index != 0 {
+			fmt.Fprintln(certPEM)
+		}
+		err = pem.Encode(certPEM, &pem.Block{
+			Type:  "CERTIFICATE",
+			Bytes: certDER,
+		})
```

Comment:
You should actually check that this is a certificate. Here

Reply:
See below response.